rfcs/0034-system-parameters.md

```diff
@@ -0,0 +1,135 @@
+---
+feature: system-parameters
+authors:
+  - "Eric Fu"
+start_date: "2023/01/11"
+---
+
+# RFC: System Parameters
+
+## Motivation
+
+Currently, we have 3 kinds of configurations:
+
+- Command-line arguments
+- Configuration file `risingwave.toml`
+- Session parameters
+
+These can not meet the requirements for several reasons.
+
+1. **Consistency**. `risingwave.toml` is a local file, which makes it possible that multiple nodes could see different parameters and this is absolutely unacceptable. For example, to configure the remote directory for storing data, users may use the `data_directory` parameter. If some nodes use `data_directory = 'hummock_001'` and others use `data_directory = 'hummock_002'`, the stored data will be corrupted.
+2. **Mutability**. Users are unable to `SET` a config after the process started. The root cause is still the local file - changes of configs must be persisted in some cluster-level storage.
+
+This RFC proposes to introduce centralized configurations on Meta Service and data will be persisted in etcd.
+
+For the sake of clarity, in this doc, we will use the term **"parameters"** for system and session parameters, and **"config"** for command-line arguments or config files.
+
+**Note:** Previously I think this might not be an urgent issue until I realized that our Cloud platform couples tightly with more and more configurations. The later we commence, the heavier burden it will become. 
+
+## Design
+
+```mermaid
+graph LR
+    all[Does it apply to the RisingWave instance?]
+    all ==No==> process["Process Configs (Local)"]
+    all ==Yes==> system["System Parameters (on Meta)"]
+    process --> cli_args(Command-line arguments)
+    process --> conf_file("Config file (.toml)")
+    system --> sql_cmd(SQL interface: SET / SHOW)
+    system --> cloud(Web Console)
+```
+
+### System Parameters
+
+What we need is a centralized source-of-truth of cluster configurations. Obviously, the best choice for RisingWave is the Meta Service and the underlying etcd. 
```

Comment:
I don't think it's a good idea to persist the configuration to the meta store. Here're several reasons:
- Meta store is not designed to be directly manipulated by the user. If one has corrupted some configuration with `SET` and finds that the cluster does not boot anymore, he can do nothing about this.
- Maintaining persisted configuration and consistency with the `clap` definition can be really hard, especially for version compatibility (some famous HTAP database has had a lot of production accidents due to this).
- It seems not convincing to persist configuration like `data_directory`... what if the user really moves the data? Note that the cluster will immediately enter the recovery mode on startup, without giving a chance to the users to set a new `data_directory`.

Reply:
Hmmm. These problems exist indeed, but persisting parameters (instead of losing all changes on reboot) is also a must-have feature for a distribution system.

> Meta store is not designed to be directly manipulated by the user. If one has corrupted some configuration with SET and finds that the cluster does not boot anymore, he can do nothing about this.

We need to equip the operation tool with the ability to fix it. For example, support a new command to `risectl set <parameter> to <value>` to manipulate `etcd` directly.  

> Maintaining persisted configuration and consistency with the `clap` definition can be really hard, especially for version compatibility (some famous HTAP database has had a lot of production accidents due to this).

`clap` definitions must refer to the CLI arguments, right? They are unrelated to system parameters in most cases.

> It seems not convincing to persist configuration like data_directory... what if the user really moves the data? Note that the cluster will immediately enter the recovery mode on startup, without giving a chance to the users to set a new data_directory.

By "persist" I mean to be recoverable after rebooting. It can be mutable, as described in section "mutability".

Reply:
> `clap` definitions must refer to the CLI arguments, right?

Or `serde`? 🥺 If we take a config file as the initialization of these parameters, then we need to have a separate definition of its schema in the code (or we can make it a dynamic `HashMap`).

> It can be mutable, as described in section "mutability".

Yes, but it cannot be mutated without a normal running cluster or `risectl` mentioned above. As the old `data_directory` is invalid, we cannot survive the recovery phase.

> but persisting parameters (instead of losing all changes on reboot) is also a must-have feature for a distribution system.

I'm thinking if we treat all `SET`s to be temporary and just for the users to give it a try, users can always write down the desired config in a new `toml` file and reboot the cluster to apply it. 🤔

Reply:
> Or `serde`? 🥺 If we take a config file as the initialization of these parameters, then we need to have a separate definition of its schema in the code (or we can make it a dynamic HashMap).

In my mind, the system parameters are all simple strings or integers so se/deserialization won't be a big issue.

> Yes, but it cannot be mutated without a normal running cluster or risectl mentioned above. As the old data_directory is invalid, we cannot survive the recovery phase.

Perhaps it also needs that operational tool: "We need to equip the operation tool with the ability to fix it" (mentioned in previous comment)

Anyway, `data_directory` is very special in many aspects: It's required to be set; it can hardly be changed when RisingWave is running; it need to be mutated offline. Things may get a little dirty for it.

> I'm thinking if we treat all SETs to be temporary and just for the users to give it a try, users can always write down the desired config in a new toml file and reboot the cluster to apply it. 🤔 I cannot remember well but I guess this is also the Flink way?

Hmmm. Interesting. It's indeed a popular way in big data systems. 

That's basically a centralized but non-persistent config store. It makes senses, but since we had a metadata database, why not just persist it? I think in most cases this looks better (except `data_dir`, I know 😇)

Reply:
This acts kind of like a mobile App and we need to provide several mechanisms to make it work, like entering the "emergency mode" and providing fixing utilities if the cluster frequently fails, or guiding them to contact customer services for asking help. 😄

I do agree we should clarify the difference between "process" configs and "cluster" configs, but I'm still doubting the possible issues and complexities for persisting it. I feel like we are taking responsibility for ourselves that could otherwise belong to the user or DevOps. 😂

Reply:
The root problem is that our system is distributed, even the Meta Service. Therefore, when someone needs to change a system parameter, she/he must change it in multiple containers/machines. That would be a pain for both human and cloud platform, and ...... looks like a design mistake 🤪  

I admit that Flink insists to use local files and I have no idea for the reasons. 🤣

Reply:
I suddenly remembered that K8s allows to mount a config as a file: https://kubernetes.io/docs/concepts/configuration/configmap/

This looks a graceful way! Although somewhat coupled with K8s

Reply:
IMHO ConfigMaps may actually be a good solution here. I am not sure how we would make this work in the risedev environment. Btw: Under the hood CMs, like all K8s objects, are also stored in ETCD. 

> Meta store is not designed to be directly manipulated by the user. If one has corrupted some configuration with SET and finds that the cluster does not boot anymore, he can do nothing about this. 

Do you think it would  help to version the configurations? Maybe this would make it simpler to fall back to a previous version if an update to the settings broke something. ETCD supports a `--prefix` operation, so you could retrieve all values belonging to a setting independent from the timestamp like this.  

```bash 
etcdctl put some_setting_23_01_27_13_22 val1 # original value
etcdctl put some_setting_23_01_28_09_00 val1 # updated value
etcdctl get some_setting --prefix # retrieve all values
some_setting_23_01_27_13_22 val1 
some_setting_23_01_28_09_00 val1
```

Reply:
> Do you think it would help to version the configurations

It's worth a separate RFC 🫣

Reply:
Sounds good. But I believe first we would have to decide on how to store the configs, before deciding if/how we want to version them.

Reply:
+1 If we really want to persist them to the meta store, we should carefully think of a way to store them. For example, we should introduce something like the "field number" in protobuf, so that we can handle the backward compatibility more easily.

Reply:
Well, I totally agree with the centralized system-wide parameters, no matter where they are stored, for the reason that it's painful for the operator to dispatch those global vars like `data_directory` now 👍 . IMO, it can be stored in 
- the etcd as a part of the metadata because we may want to change them afterward
- a local and consistent file mounted on the meta nodes so that the meta node can be the single source of truth, as all of the other components will contact meta before they get started